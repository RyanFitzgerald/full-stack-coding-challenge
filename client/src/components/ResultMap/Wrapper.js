import styled from 'styled-components';

const Wrapper = styled.div`
  border: 1px solid #e6e6e6;
  padding: 10px;
`;

export default Wrapper;