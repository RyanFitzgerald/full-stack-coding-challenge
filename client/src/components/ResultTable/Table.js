import styled from 'styled-components';

const Table = styled.table`
  border-collapse: collapse;
  border-spacing: 0;
  width: 100%;
`;

export default Table;