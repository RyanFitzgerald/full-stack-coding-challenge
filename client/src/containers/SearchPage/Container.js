import styled from 'styled-components';
import OriginalContainer from '../../components/Container';


const Container = styled(OriginalContainer)`
  padding: 50px 15px;
`;

export default Container;