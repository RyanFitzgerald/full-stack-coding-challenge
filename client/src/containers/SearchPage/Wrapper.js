import styled from 'styled-components';

const Wrapper = styled.div`
  padding: 50px 0;
`;

export default Wrapper;