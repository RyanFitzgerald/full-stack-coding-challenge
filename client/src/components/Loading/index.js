import React from 'react';

import LoadingGif from '../../static/loading.gif';
import Wrapper from './Wrapper';

function Loading() {
  return (
    <Wrapper>
      <img src={LoadingGif} alt="Loading"/>
    </Wrapper>
  );
}

export default Loading;