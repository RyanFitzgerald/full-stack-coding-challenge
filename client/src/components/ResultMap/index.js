import React from 'react';
import { Marker } from 'react-google-maps';
import { mapAPIKey } from '../../config';

import Map from '../Map';
import Wrapper from './Wrapper';

function renderMarker(address, key) {
  return (
    <Marker 
      key={key}
      title={address.formattedAddress}
      position={{lat: address.location.coordinates[1], lng: address.location.coordinates[0]}} 
    />
  );
}

function ResultMap({ data }) {
  return (
    <Wrapper>
      <Map
        googleMapURL={`https://maps.googleapis.com/maps/api/js?v=3.exp&libraries=geometry,drawing,places&key=${mapAPIKey}`}
        loadingElement={<div style={{ height: `100%` }} />}
        containerElement={<div style={{ height: `400px` }} />}
        mapElement={<div style={{ height: `100%` }} />}
      >
        {data.map((address, key) => renderMarker(address, key))}
      </Map>
    </Wrapper>
  );
}

export default ResultMap;