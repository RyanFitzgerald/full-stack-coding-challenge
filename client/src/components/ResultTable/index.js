import React from 'react';

import Table from './Table';
import Th from './Th';
import Tr from './Tr';
import Td from './Td';

function ResultTable() {
  return (
    <Table>
      <tr>
        <Th>Name</Th>
        <Th>Telephone</Th>
      </tr>
      <Tr>
        <Td>Bill Gates</Td>
        <Td>55577854</Td>
      </Tr>
      <Tr>
        <Td>Bill Gates</Td>
        <Td>55577854</Td>
      </Tr>
      <Tr>
        <Td>Bill Gates</Td>
        <Td>55577854</Td>
      </Tr>
    </Table>
  );
}

export default ResultTable;