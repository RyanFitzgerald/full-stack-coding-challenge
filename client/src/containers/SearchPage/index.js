import React from 'react';
import Helmet from 'react-helmet';
import { toast } from 'react-toastify';
import { getRequest } from '../../utils/api';

// Import components
import Loading from '../../components/Loading';
import SubHeader from '../../components/SubHeader';
import H1 from '../../components/H1';
import H2 from '../../components/H2';
import P from '../../components/P';
import Search from '../../components/Search';
import ResultTable from '../../components/ResultTable';
import ResultMap from '../../components/ResultMap';
import Button from '../../components/Button';
import Container from './Container';

class SearchPage extends React.Component {
  constructor(props) {
    super(props);

    // Bind functions
    this.handleTypeChange = this.handleTypeChange.bind(this);
    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleLoadButton = this.handleLoadButton.bind(this);
    this.handleMarkerClick = this.handleMarkerClick.bind(this);
    this.getAddresses = this.getAddresses.bind(this);
    this.renderResults = this.renderResults.bind(this);
    this.renderLoadButton = this.renderLoadButton.bind(this);

    // Initial state
    this.state = {
      type: 'List',
      query: '',
      sets: 1,
      openInfoWindow: false,
      data: false
    };
  }

  componentDidMount() {
    const { query } = this.state;
    this.getAddresses(query);
  }

  // Handle changes of the result type
  handleTypeChange(type) {
    this.setState({ type });
  }

  // Handle search input changes
  handleQueryChange(query) {
    this.getAddresses(query);
    this.setState({ query });
  }

  // Handle search input changes
  handleLoadButton() {
    this.setState((prevState) => ({
      sets: prevState.sets + 1
    }));
  }

  handleMarkerClick(id) {
    this.setState({
      openInfoWindow: id
    });
  }

  getAddresses(query) {
    getRequest(`/api/v1/addresses?q=${query}`).then(data => {
      if (typeof data === 'undefined' || data.status === 500 || data.error) {
        toast.error("An error has occured, please try again.");
        return
      }
      this.setState({ data });
    });
  }

  // Render results based on type chosen
  renderResults(type, sets, data) {
    // Destructure state
    const { openInfoWindow } = this.state;

    // If there is no data, show loading
    if (!data) {
      return (
        <Loading/>
      );
    }

    // If type is Map, render the map
    if (type === 'Map') {
      return (
        <ResultMap openInfoWindow={openInfoWindow} onMarkerClick={this.handleMarkerClick} data={data}/>
      );
    }

    // Otherwise, render the table
    return (
      <ResultTable data={data} sets={sets}/>
    );
  }

  // Render the "Load More" button if needed (i.e. more data and type 'List')
  renderLoadButton(type, sets, dataLength) {
    // Only show if not a map and you haven't reached the end
    if (type !== 'Map' && (sets * 20) <= dataLength) {
      return (
        <Button onClick={this.handleLoadButton}>Load More</Button>
      );
    }
  }

  render() {
    // Destructure State
    const { type, query, sets, data } = this.state;

    return (
      <div>
        <Helmet>
          <title>Address Search - Full-stack Coding Challenge - Ambyint</title>
        </Helmet>
        <SubHeader>
          <H1>Address Search</H1>
          <P>Use the search form below to look up addresses</P>
          <Search 
            activeType={type}
            query={query}
            onQueryChange={this.handleQueryChange}
            onTypeChange={this.handleTypeChange}
          />
        </SubHeader>
        <Container>
          <H2>{data.length} Result(s)</H2>
          {this.renderResults(type, sets, data)}
          {this.renderLoadButton(type, sets, data.length)}
        </Container>
      </div>
    );
  }
}

export default SearchPage;