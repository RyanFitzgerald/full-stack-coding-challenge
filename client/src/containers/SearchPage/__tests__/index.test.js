import React from 'react';

import SearchPage from '../index';
import Loading from '../../../components/Loading';
import ResultTable from '../../../components/ResultTable';
import ResultMap from '../../../components/ResultMap';

describe('<SearchPage />', () => {
  it('should match the snapshot', () => {
    const renderedComponent = renderer.create(
      <SearchPage/>
    ).toJSON();
    
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should render the loading component', () => {
    const renderedComponent = mount(
      <SearchPage/>
    );

    expect(renderedComponent.find(Loading).length).toEqual(1);
  });

  it('should render the result table', () => {
    const data = [{
      _id: '5b428620b4a1f026a4c6058b',
      city: 'Lenexa',
      country: 'United States',
      dateAdded: '2018-07-08T21:46:08.417Z',
      formattedAddress: '15481 W 110th St, Lenexa, KS 66219, USA',
      location: {
        coordinates: [-94.7655029, 38.928542],
        type: 'Point'
      },
      postalCode: '66219',
      state: 'Kansas',
      street: 'West 110th Street',
      streetNumber: '15481'
    }];

    const renderedComponent = mount(
      <SearchPage/>
    );

    renderedComponent.setState({ data });

    expect(renderedComponent.find(ResultTable).length).toEqual(1);
  });

  it('should render the result map', () => {
    const data = [{
      _id: '5b428620b4a1f026a4c6058b',
      city: 'Lenexa',
      country: 'United States',
      dateAdded: '2018-07-08T21:46:08.417Z',
      formattedAddress: '15481 W 110th St, Lenexa, KS 66219, USA',
      location: {
        coordinates: [-94.7655029, 38.928542],
        type: 'Point'
      },
      postalCode: '66219',
      state: 'Kansas',
      street: 'West 110th Street',
      streetNumber: '15481'
    }];

    const renderedComponent = shallow(
      <SearchPage/>
    );

    renderedComponent.setState({ data, type: 'Map' });
    
    expect(renderedComponent.find(ResultMap).length).toEqual(1);
  });
});