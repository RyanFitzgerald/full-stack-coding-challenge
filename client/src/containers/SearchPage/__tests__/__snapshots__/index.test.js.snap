// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage /> should match the snapshot 1`] = `
.c12 {
  margin: 30px 0;
  text-align: center;
}

.c0 {
  background: #f5f5f5;
  background-image: url(bg.png);
  border-bottom: 1px solid #e6e6e6;
  padding: 100px 0;
  text-align: center;
}

.c1 {
  margin: 0 auto;
  max-width: 960px;
  padding: 15px;
}

.c1:after {
  content: '';
  display: table;
  clear: both;
}

.c2 {
  color: #373958;
  font-size: 2.5em;
  line-height: 1.25;
  margin: 0;
}

.c11 {
  color: #373958;
  font-size: 2em;
  line-height: 1.25;
  margin: 15px 0;
  text-align: center;
}

.c3 {
  margin: 0;
}

.c4 {
  margin-top: 30px;
  position: relative;
}

.c5 {
  bottom: 0;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  width: 50px;
}

.c5 svg {
  margin-top: 12px;
}

.c6 {
  border: 1px solid #e6e6e6;
  border-radius: 4px;
  box-shadow: 0 1px 10px rgba(67,66,97,0.5);
  outline: none;
  padding: 15px 15px 15px 50px;
  width: 100%;
}

.c8 {
  background: #e5e5e5;
  border: 1px solid #e6e6e6;
  border-bottom-right-radius: 4px;
  border-top-right-radius: 4px;
  bottom: 0;
  color: #696d80;
  cursor: pointer;
  padding: 0;
  position: absolute;
  right: 0;
  top: 0;
  -webkit-transition: 0.25s ease all;
  transition: 0.25s ease all;
  width: 100px;
}

.c8:hover {
  background: #e5e5e5;
}

.c9 {
  background: #fafafa;
  border: 1px solid #e6e6e6;
  border-bottom-right-radius: 4px;
  border-top-right-radius: 4px;
  bottom: 0;
  color: #696d80;
  cursor: pointer;
  padding: 0;
  position: absolute;
  right: 0;
  top: 0;
  -webkit-transition: 0.25s ease all;
  transition: 0.25s ease all;
  width: 100px;
}

.c9:hover {
  background: #e5e5e5;
}

.c7 {
  border-bottom-right-radius: 0;
  border-right: 0;
  border-top-right-radius: 0;
  right: 100px;
}

.c10 {
  padding: 50px 15px;
}

@media only screen and (max-width:480px) {
  .c6 {
    padding-right: 15px;
  }
}

@media only screen and (max-width:480px) {
  .c8 {
    bottom: -100%;
    height: 40px;
    left: auto;
    right: 0;
    top: auto;
    width: 50%;
  }
}

@media only screen and (max-width:480px) {
  .c9 {
    bottom: -100%;
    height: 40px;
    left: auto;
    right: 0;
    top: auto;
    width: 50%;
  }
}

@media only screen and (max-width:480px) {
  .c7 {
    left: 0;
    right: auto;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
    >
      <h1
        className="c2"
      >
        Address Search
      </h1>
      <p
        className="c3"
      >
        Use the search form below to look up addresses
      </p>
      <div
        className="c4"
      >
        <div
          className="c5"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="11"
              cy="11"
              r="8"
            />
            <line
              x1="21"
              x2="16.65"
              y1="21"
              y2="16.65"
            />
          </svg>
        </div>
        <input
          className="c6"
          defaultValue=""
          name="Search"
          onChange={[Function]}
          placeholder="Enter Search Term"
          type="text"
        />
        <button
          className="c7 c8"
          onClick={[Function]}
        >
          List View
        </button>
        <button
          className="c9"
          onClick={[Function]}
        >
          Map View
        </button>
      </div>
    </div>
  </div>
  <div
    className="c10 c1"
  >
    <h2
      className="c11"
    >
       Result(s)
    </h2>
    <div
      className="c12"
    >
      <img
        alt="Loading"
        src="loading.gif"
      />
    </div>
  </div>
</div>
`;
